In this case, the Automobile Data set is an online source, and it is in a CSV (comma separated value) format.

* Data source: https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
* Data type: csv

In [1]:
#import libraries
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
#load the data into a dataframe
filepath = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'
df = pd.read_csv(filepath, header=None)

In [3]:
df.head()

0    1            2    3    4     5            6    7      8     9   ...  \
0   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
1   3    ?  alfa-romero  gas  std   two  convertible  rwd  front  88.6  ...   
2   1    ?  alfa-romero  gas  std   two    hatchback  rwd  front  94.5  ...   
3   2  164         audi  gas  std  four        sedan  fwd  front  99.8  ...   
4   2  164         audi  gas  std  four        sedan  4wd  front  99.4  ...   

    16    17    18    19    20   21    22  23  24     25  
0  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
4  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  

[5 rows x 26 columns]

In [4]:
#add header to the dataframe
header = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style","drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type","num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower","peak-rpm","city-mpg","highway-mpg","price"]
df.columns = header

#**Identify missing values**
###**Convert "?" to NaN**
In the car data set, missing data comes with the question mark "?". We replace "?" with NaN (Not a Number), Python's default missing value marker for reasons of computational speed and convenience.

In [5]:
#replace the "?" symbol with NaN so the dropna() can remove the missing values
df.replace("?", np.nan, inplace = True)
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [6]:
#count missing values in every column
missing_data = df.isnull()
for col in missing_data.columns:
  print(missing_data[col].value_counts())
  print(" ")

symboling
False    205
Name: count, dtype: int64
 
normalized-losses
False    164
True      41
Name: count, dtype: int64
 
make
False    205
Name: count, dtype: int64
 
fuel-type
False    205
Name: count, dtype: int64
 
aspiration
False    205
Name: count, dtype: int64
 
num-of-doors
False    203
True       2
Name: count, dtype: int64
 
body-style
False    205
Name: count, dtype: int64
 
drive-wheels
False    205
Name: count, dtype: int64
 
engine-location
False    205
Name: count, dtype: int64
 
wheel-base
False    205
Name: count, dtype: int64
 
length
False    205
Name: count, dtype: int64
 
width
False    205
Name: count, dtype: int64
 
height
False    205
Name: count, dtype: int64
 
curb-weight
False    205
Name: count, dtype: int64
 
engine-type
False    205
Name: count, dtype: int64
 
num-of-cylinders
False    205
Name: count, dtype: int64
 
engine-size
False    205
Name: count, dtype: int64
 
fuel-system
False    205
Name: count, dtype: int64
 
bore
False    201
True       4
Na

From the summary above we can see that each column contains 205 entries of data. It can be noted than 7 columns contain missing data:
- Normalized-losses: 41
- Num-of-doors: 2
- Bore: 4
- Stroke: 4
- Horse-power: 2
- Peak-rpm: 2
- Price: 4

##**Dealing with missing values**

###**Droping rows**
Since we want to predict prices, we can't use entries without price data for prediction.
For the column "price", we'll delete the rows for missing data

In [7]:
#drop entire rows with missing values along the column price
df.dropna(subset=["price"], axis=0, inplace=True)

###**Replace by mean**
We replace missing values from the columns "normalized-losses", "bore", "stroke", "horse-power", "peak-rpm".

In [8]:
#calculate the mean values for the above columns
#normalized-loss mean
norm_loss_avg = df["normalized-losses"].astype("float").mean(axis=0)
#bore mean
bore_avg = df['bore'].astype('float').mean(axis=0)
#stroke mean
stroke_avg = df['stroke'].astype('float').mean(axis=0)
#horsepower mean
horsepow_avg = df['horsepower'].astype('float').mean(axis=0)
#peak rpm mean
peakrpm_avg = df['peak-rpm'].astype('float').mean(axis=0)

print("Normalized-losses avg: ", norm_loss_avg)
print("Bore avg: ", bore_avg)
print("Stroke avg: ", stroke_avg)
print("Horsepower avg: ", horsepow_avg)
print("Peak rpm avg: ", peakrpm_avg)

Normalized-losses avg:  122.0
Bore avg:  3.330710659898477
Stroke avg:  3.256903553299492
Horsepower avg:  103.39698492462311
Peak rpm avg:  5117.587939698493


In [9]:
#replace "NaN" with the mean values above
#normalized-losses
df["normalized-losses"].replace(np.nan, norm_loss_avg, inplace=True)
#bore
df['bore'].replace(np.nan, bore_avg, inplace=True)
#stroke
df['stroke'].replace(np.nan, stroke_avg, inplace=True)
#horsepower
df['horsepower'].replace(np.nan, horsepow_avg, inplace=True)
#peak rpm
df['peak-rpm'].replace(np.nan, peakrpm_avg, inplace=True)

###**Replace by frequency**
For the column "num-of-doors" we check for the values present and replace the missing values with the most frequently occuring value.

In [10]:
#check the values present in the column "num-of-doors"
df['num-of-doors'].value_counts()

num-of-doors
four    113
two      86
Name: count, dtype: int64

In [11]:
#replace the missing data with the most common value i.e four
comm_num_doors = df['num-of-doors'].value_counts().idxmax()
df['num-of-doors'].replace(np.nan, comm_num_doors, inplace=True)

In [12]:
#count missing values in every column after handling missing data
missing_data = df.isnull()
for col in missing_data.columns:
  print(missing_data[col].value_counts())
  print(" ")

symboling
False    201
Name: count, dtype: int64
 
normalized-losses
False    201
Name: count, dtype: int64
 
make
False    201
Name: count, dtype: int64
 
fuel-type
False    201
Name: count, dtype: int64
 
aspiration
False    201
Name: count, dtype: int64
 
num-of-doors
False    201
Name: count, dtype: int64
 
body-style
False    201
Name: count, dtype: int64
 
drive-wheels
False    201
Name: count, dtype: int64
 
engine-location
False    201
Name: count, dtype: int64
 
wheel-base
False    201
Name: count, dtype: int64
 
length
False    201
Name: count, dtype: int64
 
width
False    201
Name: count, dtype: int64
 
height
False    201
Name: count, dtype: int64
 
curb-weight
False    201
Name: count, dtype: int64
 
engine-type
False    201
Name: count, dtype: int64
 
num-of-cylinders
False    201
Name: count, dtype: int64
 
engine-size
False    201
Name: count, dtype: int64
 
fuel-system
False    201
Name: count, dtype: int64
 
bore
False    201
Name: count, dtype: int64
 
stroke
False 

Now our data is free of missing data

##**Correct data format**


In [13]:
#list data format for all columns
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

From the summary above we can see that some columns are not of the correct data types. Numerical variables should be type 'int' or 'float' while string variable should have type 'object'.



In [14]:
#correct data types to the right format
df[['price']] = df[['price']].astype('float')
df[['peak-rpm']] = df[['peak-rpm']].astype('float')
df[['horsepower']] = df[['horsepower']].astype('float')
df[['stroke']] = df[['stroke']].astype('float')
df[['bore']] = df[['bore']].astype('float')
df[['normalized-losses']] = df[['normalized-losses']].astype('int')

In [15]:
#list data formats for each column after correction
df.dtypes

symboling              int64
normalized-losses      int64
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                 float64
stroke               float64
compression-ratio    float64
horsepower           float64
peak-rpm             float64
city-mpg               int64
highway-mpg            int64
price                float64
dtype: object

In [ ]:
df.to_csv('automobile.csv', index=False)